test/java/seedu/addressbook/commands/ViewCommandTest.java

```diff
@@ -0,0 +1,157 @@
+
+    @Before
+    public void setUp() throws Exception {
+        emptyAddressBook = TestUtil.createAddressBook();
+
+        TypicalPersons td = new TypicalPersons();
+        addressBook = td.getTypicalAddressBook();
+        listWithAll = Arrays.asList(td.getTypicalPersons());
+        listWithSome = Arrays.asList(td.amy, td.candy, td.dan);
+    }
+
+    @Test
+    public void viewCommand_invalidIndex_returnsInvalidIndexMessage() {
+        // empty addressbook
+        assertViewErrorInvalidIndex(emptyAddressBook, emptyDisplayList, 1);
+
+        // non-empty addressbook
+        assertViewErrorInvalidIndex(addressBook, listWithAll, -1);
+        assertViewErrorInvalidIndex(addressBook, listWithAll, 0);
+        assertViewErrorInvalidIndex(addressBook, listWithAll, listWithAll.size() + 1);
+    }
+
+    @Test
+    public void viewCommand_personNotInAddressBook_returnsPersonNotInAddressBookMessage() throws Exception {
+        // generate person not in addressbook, add to displayList
+        ReadOnlyPerson someone = new Person(new Name("me"),
+                                            new Phone("123", true),
+                                            new Email("some@hey.go", true),
+                                            new Address("nus", false),
+                                            new UniqueTagList(Collections.emptySet()));
+        listWithAll = new ArrayList<ReadOnlyPerson>(listWithAll);
+        listWithAll.add(someone);
+
+        // empty addressbook
+        assertViewErrorPersonNotInAddressBook(emptyAddressBook, listWithAll, 1);
+
+        // non-empty addressbook
+        assertViewErrorPersonNotInAddressBook(addressBook, listWithAll, listWithAll.size());
+    }
+
+    @Test
+    public void viewCommand_validIndex_returnsPersonDetails() {
+        // person with no private information
+        assertViewSuccess(addressBook, listWithAll, 1);
+
+        // person with some private information
+        assertViewSuccess(addressBook, listWithAll, 2);
+
+        // person with all private information
+        assertViewSuccess(addressBook, listWithAll, 4);
+
+        // addressbook has more people than displayList
+        // This can happen when view from filtered list caused by some commands(eg. FindCommand)
+        assertViewSuccess(addressBook, listWithSome, 1);
+    }
+
+    private Command generateViewCommand(AddressBook addressBook, List<ReadOnlyPerson> displayList, int index) {
+        ViewCommand command = new ViewCommand(index);
+        command.setData(addressBook, displayList);
+
+        return command;
+    }
+
+    private Command generateViewAllCommand(AddressBook addressBook, List<ReadOnlyPerson> displayList, int index) {
+        Command command = new ViewAllCommand(index);
+        command.setData(addressBook, displayList);
+
+        return command;
+    }
+
+    /**
+     * Asserts that the details of the person at specific index can be successfully retrieved
+     * and displayed.
```

Comment:
`Asserts that both a ViewCommand and a ViewAllCommand can retrieve from the {@code addressBook} details of the person at the given {@code index} in the given {@code list}.`